src/company_gui.cpp

```diff
@@ -1827,267 +1845,217 @@ struct CompanyInfrastructureWindow : Window
-				max_val = std::max(max_val, c->infrastructure.station);
-				max_cost = std::max(max_cost, StationMaintenanceCost(c->infrastructure.station));
-				max_val = std::max(max_val, c->infrastructure.airport);
-				max_cost = std::max(max_cost, AirportMaintenanceCost(c->index));
-
-				SetDParamMaxValue(0, max_val);
-				uint count_width = GetStringBoundingBox(STR_JUST_COMMA).width + WidgetDimensions::scaled.hsep_indent; // Reserve some wiggle room
-
-				if (_settings_game.economy.infrastructure_maintenance) {
-					StringID str_total = TimerGameEconomy::UsingWallclockUnits() ? STR_COMPANY_INFRASTRUCTURE_VIEW_TOTAL_PERIOD : STR_COMPANY_INFRASTRUCTURE_VIEW_TOTAL_YEAR;
-					SetDParamMaxValue(0, this->GetTotalMaintenanceCost() * 12); // Convert to per year
-					this->total_width = GetStringBoundingBox(str_total).width + WidgetDimensions::scaled.hsep_indent * 2;
-					size.width = std::max(size.width, this->total_width);
-
-					SetDParamMaxValue(0, max_cost * 12); // Convert to per year
-					count_width += std::max(this->total_width, GetStringBoundingBox(str_total).width);
-				}
 
-				size.width = std::max(size.width, count_width);
+		/* air-ways. */
+		infrastructurelist.emplace_back(CIW_ENTRY_HEADER, STR_COMPANY_INFRASTRUCTURE_VIEW_STATION_SECT);
 
-				/* Set height of the total line. */
-				if (widget == WID_CI_TOTAL) {
-					size.height = _settings_game.economy.infrastructure_maintenance ? std::max<uint>(size.height, WidgetDimensions::scaled.vsep_normal + GetCharacterHeight(FS_NORMAL)) : 0;
-				}
+		maintenance_cost        = StationMaintenanceCost(c->infrastructure.station);
+		total_maintenance_cost += maintenance_cost;
+
+		infrastructurelist.emplace_back(CIW_ENTRY_LABEL_VALUES, STR_COMPANY_INFRASTRUCTURE_VIEW_STATIONS, c->infrastructure.station, maintenance_cost);
+
+		maintenance_cost        = AirportMaintenanceCost(c->index);
+		total_maintenance_cost += maintenance_cost;
+
+		infrastructurelist.emplace_back(CIW_ENTRY_LABEL_VALUES, STR_COMPANY_INFRASTRUCTURE_VIEW_AIRPORTS, c->infrastructure.airport, maintenance_cost);
+
+
+		/* total maintenance cost. */
+		infrastructurelist.emplace_back(CIW_ENTRY_TOTAL_COST, STR_COMPANY_VIEW_INFRASTRUCTURE_NONE, 0, total_maintenance_cost);
+
+
+		this->vscroll->SetCount(this->infrastructurelist.size());  // Update scrollbar as well.
+	}
+
+	void SetStringParameters(WidgetID widget) const override
+	{
+		switch (widget) {
+			case WID_CI_CAPTION:
+				SetDParam(0, (CompanyID)this->window_number);
 				break;
-			}
 		}
 	}
 
-	/**
-	 * Helper for drawing the counts line.
-	 * @param r            The bounds to draw in.
-	 * @param y            The y position to draw at.
-	 * @param count        The count to show on this line.
-	 * @param monthly_cost The monthly costs.
-	 */
-	void DrawCountLine(const Rect &r, int &y, int count, Money monthly_cost) const
+	inline int GetLineHeight() const
+	{
+		return GetCharacterHeight(FS_NORMAL) + WidgetDimensions::unscaled.vsep_normal / 2;
```

Comment:
Using / 2 on a standard scaled WidgetDimension seems wrong to me. The original list did not have this padding and used normal character height spacing. I don't think that needs to change.